### PR TITLE
add oci platform filter

### PR DESCRIPTION
**Pull request by jschicktanz**
**What this PR does / why we need it**:
add oci platform filter

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```other operator
- add oci platform filter
```

### CODE REVIEW

File: oci/platform.py [hunk 1]
```diff
@@ -7,6 +8,56 @@
...
+    JS = 'js'
+    LINUX = 'linux'
+    NETBSD = 'netbsd'
+    OPENBSD = 'openbsd'
+    PLAN9 = 'plan9'
+    SOLARIS = 'solaris'
+    WINDOWS = 'windows'
+
+    @classmethod
+    def contains_value(cls, value: str):
```
Comment (ccwienk) [inline oci/platform.py:32]: rather overwrite __contains__
as an (more pythonic) alternative, one could resort to leveraging enum.Enum's autogenerated initialiser's semantics

-> `OperatingSystem('does-not-exist')` will raise `ValueError`

File: oci/platform.py [hunk 2]
```diff
@@ -7,6 +8,56 @@
...
+    MIPS = 'mips'
+    MIPSLE = 'mipsle'
+    MIPS64 = 'mips64'
+    MIPS64LE = 'mips64le'
+    PPC64le = 'ppc64le'
+    RISCV64 = 'riscv64'
+    S390X = 's390x'
+
+    @classmethod
+    def contains_value(cls, value: str):
```
Comment (ccwienk) [inline oci/platform.py:57]: ditto

File: oci/platform.py [hunk 3]
```diff
@@ -160,3 +211,57 @@ def iter_platforms(
...
 
         yield (sub_img_ref, platform)
+
+
+class PlatformFilter:
+    @staticmethod
+    def create(
+        included_platforms: typing.List[str],
+    ) -> typing.Callable[[om.OciPlatform], bool]:
+        matchers = []
```
Comment (ccwienk) [inline oci/platform.py:221]: nit: use list-comprehension

File: oci/platform.py [hunk 4]
```diff
@@ -160,3 +211,57 @@ def iter_platforms(
...
+                if PlatformFilter._match(m, platform_to_match):
+                    return True
+
+            return False
+
+        return filter
+
+    @staticmethod
+    def _parse_expr(platform_expr: str) -> dict:
+        splitted = platform_expr.split('/')
```
Comment (ccwienk) [inline oci/platform.py:236]: nit: split, split, split

File: oci/platform.py [hunk 5]
```diff
@@ -160,3 +211,57 @@ def iter_platforms(
...
+        if len(splitted) < 2 or len(splitted) > 3:
```
Comment (ccwienk) [inline oci/platform.py:237]: rather use separate if-clauses for both variants - or at least output actual length in error-msg to make debugging easier

Comment (ccwienk) [reply oci/platform.py:237]: you might want to consider using structural pattern matching (https://peps.python.org/pep-0622/)

File: oci/platform.py [hunk 6]
```diff
@@ -160,3 +211,57 @@ def iter_platforms(
...
+            raise ValueError(f'invalid oci platform expression {platform_expr=}.'
+                              ' expression must have the format os/architecture[/variant]')
+
+        os = splitted[0]
+        if os != '*' and not OperatingSystem.contains_value(os):
+            raise ValueError(f'invalid os in oci platform expression {platform_expr=}.'
+                             f' allowed values are {["*"] + [o.value for o in OperatingSystem]}')
+
+        architecture = splitted[1]
+        if architecture != '*' and not Architecture.contains_value(architecture):
```
Comment (ccwienk) [inline oci/platform.py:247]: as mentioned above: simply use Archecture(value) and let ValueError propagate

File: oci/platform.py [hunk 7]
```diff
@@ -160,3 +211,57 @@ def iter_platforms(
...
+            variant = splitted[2]
+
+        return {
+            'os': os,
+            'architecture': architecture,
+            'variant': variant,
+        }
+
+    @staticmethod
+    def _match(m: dict, p: om.OciPlatform) -> bool:
```
Comment (ccwienk) [inline oci/platform.py:262]: there is __eq__ for that

**Review by ccwienk — CHANGES_REQUESTED** [review]
generally speaking, I think I would prefer to keep most of the platform-handling code w/ the using code, and keep semantic focus of `oci` package on both model classes for (de-)serialisation of oci-manifests (+ sub-datastructures), as well as interactions w/ oci-api. Normalisation seems like a good idea (however, I am uncertain whether we should actually be strict and specify allowed platforms/architectures - this is something we do not need to restrict, and could resort to just see thos IDs as strings). As for filtering, I think this would fit better into user's code.

this will allow us to keep oci-package and using "ctt" more losely coupled

**Comment by jschicktanz** [comment]
coding was moved to https://github.com/gardener/cnudie-transport-tool/pull/488

**Pull request closed without merging**